support-services/src/main/scala/com/gu/zuora/ZuoraGiftService.scala

```diff
@@ -44,8 +50,73 @@ class ZuoraGiftService(val config: ZuoraConfig, client: FutureHttpClient)(implic
+          SafeLogger.info(s"Successfully updated gifteeIdentityId on Digital Subscription gift for user $gifteeIdentityId")
+          Success(successfulResponse)
+        case Success(failedResponse) =>
+          // Treat a ZuoraResponse where success = false as a failure
+          val responseMessage = failedResponse.errorMessage.getOrElse("unknown")
+          val errorMessage = s"Failed to update gifteeIdentityId on Digital Subscription gift for user $gifteeIdentityId. Error was ${responseMessage}"
+          Failure(new RuntimeException(errorMessage))
+        case Failure(originalError) =>
+          val errorMessage = s"Failed to update gifteeIdentityId on Digital Subscription gift for user $gifteeIdentityId."
+          Failure(new RuntimeException(errorMessage, originalError))
+      }
+  }
+
+  def setupRevenueRecognition(
+    subscription: Subscription,
+    termDates: TermDates
+  ) = {
+    //This doc describes what this method is doing: https://docs.google.com/document/d/1yNeaR2l1Ss_unXygntuntmRX-GicDelryuK25vmqToc/edit
+    val response = (for {
+      ratePlan <- OptionT.fromOption(subscription.ratePlans.headOption)
+      ratePlanChargeId <- OptionT.fromOption(ratePlan.ratePlanCharges.headOption.map(_.id))
+      revenueSchedule <- OptionT.liftF(getRevenueSchedule(ratePlanChargeId))
+      revenueScheduleNumber <- OptionT.fromOption(revenueSchedule.revenueSchedules.headOption.map(_.number))
+      successOrFailureResponse <- OptionT.liftF(distributeRevenueByStartAndEndDates(revenueScheduleNumber, termDates))
+    } yield successOrFailureResponse).value
+
+    val subscriptionNumber = subscription.subscriptionNumber
+    // For the revenue recognition we want to continue even if there is a failure because by this point
+    // the redemption will have succeeded so we need to let the user know about that and then sort out the
+    // revenue recognition separately
+    response.transform{
+      case Success(Some(response)) if response.success =>
+        SafeLogger.info(s"Successfully set up revenue recognition for Digital Subscription gift $subscriptionNumber")
+        Success(())
+      case Success(Some(response)) =>
+        val message = response.errorMessage.getOrElse("unknown")
+        val errorMessage = scrub"Failed to set up revenue recognition for Digital Subscription gift $subscriptionNumber. Error was ${message}"
+        SafeLogger.error(errorMessage) //TODO: Add an alarm
+        Success(())
+      case Failure(originalError) =>
+        val errorMessage = scrub"Failed to set up revenue recognition for Digital Subscription gift $subscriptionNumber."
+        SafeLogger.error(errorMessage, originalError) //TODO: Add an alarm
+        Success(())
+      case _ =>
+        val errorMessage = scrub"Failed to set up revenue recognition for Digital Subscription gift $subscriptionNumber."
+        SafeLogger.error(errorMessage) //TODO: Add an alarm
+        Success(())
```

Comment:
I know it will be slightly less easy to read the message, but I feel like the error handling outweighs the actual logic.  What do you think?
```suggestion
      case other =>
        val errorMessage = scrub"Failed to set up revenue recognition for Digital Subscription gift $subscriptionNumber. Error was $other"
        SafeLogger.error(errorMessage) //TODO: Add an alarm
        Success(())
```

Reply:
Yes I think you're probably right